src/osdc/Journaler.cc

```diff
@@ -1292,7 +1292,7 @@ void Journaler::wait_for_readable(Context *onreadable)
   }
 
   ceph_assert(on_readable == 0);
-  if (!readable) {
+  if (!readable && read_pos < write_pos) {
```

Comment:
Could you elaborate about this fix? Under what case is `readable` unset but `read_pos < write_pos`?

Reply:
In the last few lines of function 'Journaler::_finish_read', if condition 'got_any && !was_readable && readable' or condition 'read_pos == write_pos' is true, and there is a waiter(variable 'on_readable' has been set) too, it will be notified. So the code 'readable_waiter.wait()' in function 'MDLog::_replay_thread' is waiting for the entry to be really readable or the end of journal('read_pos == write_pos') after function 'Journaler::wait_for_readable' is called. So the function 'C_SaferCond::wait' called later need to return immediately if one of these two conditions is already true before it is called. So both conditions need to be checked in function 'Journaler::wait_for_readable', and then make the function 'C_SaferCond::wait' called later return immediately if one of them is true. We can see the first condition(variable 'readable') is checked, and variable 'onreadable' will be put into the queue of finisher if the condition is true, that's right. But we don't see the second condition being checked, it seems to be missing. So I Add this condition check so that variable 'onreadable' is also put into the queue of finisher if 'read_pos == write_pos'.
We also need to see the impact on other callers after modifying function 'Journaler::wait_for_readable':
(1) MDLog::_reformat_journal
There is the same logic(while...wait_for_readable...wait) in this function and the code comments also say the logic here is borrowed from replay_thread, so we will also solve the same problem here.
(2) PurgeQueue::_recover
There is the similar logic here in this function(it needs to be called when readable or the end of journal is reached), so it may has the same problem and we can solve it.
(3) PurgeQueue::_consume
The function 'wait_for_readable' be called here is waiting for readable even if 'read_pos == write_pos' is already true, and then call function 'PurgeQueue::_consume' again. So our changes to function 'wait_for_readable' will affect here so that function 'PurgeQueue::_consume' will be called immediately instead of waiting when condition 'read_pos == write_pos' is true. But I think it is not necessary to keep waiting for readable here when condition 'read_pos == write_pos' is true, because variable 'write_pos' needs to be increased when readable again and it can only be done by code 'journaler.append_entry(bl)' in function 'PurgeQueue::push', and then function 'PurgeQueue::_consume' will be called too. So add a condition check in function 'PurgeQueue::_consume' to let it return directly if 'read_pos == write_pos'. And it may need another PR, or put it here too.

Reply:
>So I Add this condition check so that variable 'onreadable' is also put into the queue of finisher if 'read_pos == write_pos'.

`Journaler::_assimilate_prefetch()` will notify waiter when `read_pos == write_pos`. From `MDLog::_replay_thread()`:

```
    while (!journaler->is_readable() &&
           journaler->get_read_pos() < journaler->get_write_pos() &&
           !journaler->get_error()) {
      C_SaferCond readable_waiter;
      journaler->wait_for_readable(&readable_waiter);
      r = readable_waiter.wait();
    }
```

`->is_readable()` will start a prefetch which in turn will call `_assimilate_prefetch()`. How is leading to a infinite wait?

Reply:
When function 'Journaler::_assimilate_prefetch' completes before function 'Journaler::wait_for_readable' is called, although variable  'on_readable' will still be set in function 'Journaler::wait_for_readable' , but it's too late. And then the code 'readable_waiter.wait()' will wait for the next 'Journaler::_assimilate_prefetch', but there will be no more.

Reply:
The way PR #48281 addressed this issue is by ensuring the relevant checks are under `Journaler::lock`. Have you looked at that change?

Reply:
I've seen it and I think it may also solve this problem.
The relevant checks has been completed in the condition of the loop, but they must to be checked again under 'Journaler::lock'. So checking them in function 'Journaler::wait_for_readable' is ok, it has the lock we need. And one of them has been checked in this function, we just need to add another.
I think all roads lead to Rome, and I like the shortest one.

Reply:
Got your point. We need the checks to be done safely under lock (PR #48281) and this change for the wake up condition to be satisfied